### PR TITLE
feat(ADN): Add German variant and season downloads

### PR DESCRIPTION
**Pull request by infanf**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

Add support for the newly launched German site. The German site is accessible from DE, AT, CH, LI, LU under animationdigitalnetwork.de (instead of .fr). Backend API is the same as on the French site.

German site provides German subtitles and dubs (currently none available).

And while I'm at it, might as well add season download support as requested in #6643 

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

### CODE REVIEW

File: yt_dlp/extractor/adn.py [hunk 1]
```diff
@@ -147,6 +158,8 @@ def _perform_login(self, username, password):
             self.report_warning(message or self._LOGIN_ERR_MESSAGE)
 
     def _real_extract(self, url):
+        if 'network.de' in url:
+            self._HEADERS['X-Target-Distribution'] = 'de'
```
Comment (pukkandan) [inline yt_dlp/extractor/adn.py:162]: Don't add this to _HEADERS. The same IE class is used for multiple URLs. Just add the header locally to relevant calls

File: yt_dlp/extractor/adn.py [hunk 2]
```diff
@@ -49,6 +49,15 @@ class ADNIE(InfoExtractor):
...
         'only_matching': True,
+    }, {
+        'url': 'https://animationdigitalnetwork.de/video/tokyo-mew-mew-new/19818-folge-1',
+        'info_dict': {
+            'id': '19818',
+            'ext': 'mp4',
+            'episode_number': 1,
+            'duration': 1429,
+        },
+        'skip': 'Only available in region (FR, ...)',
```
Comment (pukkandan) [inline yt_dlp/extractor/adn.py:60]: Did you mean

```suggestion
        'skip': 'Only available in Germany',
```

Comment (pukkandan) [inline yt_dlp/extractor/adn.py:60]: Try running the test locally after commenting out `skip`

File: yt_dlp/extractor/adn.py [hunk 3]
```diff
@@ -116,6 +125,8 @@ def _get_subtitles(self, sub_url, video_id):
 
             if sub_lang == 'vostf':
                 sub_lang = 'fr'
+            if sub_lang == 'vostde':
```
Comment (pukkandan) [inline yt_dlp/extractor/adn.py:128]:
```suggestion
            elif sub_lang == 'vostde':
```

**Comment by infanf** [comment]
The German site is still under contrsuction and some stream variants (namely mobile and sd quality) are not yet available. Will keep this as a draft until the German site is working.

File: README.md [hunk 4]
```diff
@@ -1886,6 +1886,10 @@ The following extractors use this feature:
 #### nflplusreplay
 * `type`: Type(s) of game replays to extract. Valid types are: `full_game`, `full_game_spanish`, `condensed_game` and `all_22`. You can use `all` to extract all available replay types, which is the default
 
+#### ADN
+* `language`: Which language version to extract. Valid values are `fr` and `de`. If not provided, the language is extracted from the provided URL
```
Comment (pukkandan) [inline README.md:1890]: What was wrong with the first implementation? Why is this necessary when user can simply pass the appropriate URL?

File: README.md [hunk 5]
```diff
@@ -1886,6 +1886,10 @@ The following extractors use this feature:
...
+* `season`: When downloading a show, download only the specified season. If not provided, all seasons are downloaded
```
Comment (pukkandan) [inline README.md:1891]: Just return the `season_number` in the video metadata and user can do `--match-filter`. `--extractor-args` are meant for use-cases that are otherwise unachievable

File: yt_dlp/extractor/adn.py [hunk 6]
```diff
@@ -255,3 +280,44 @@ def _real_extract(self, url):
             'average_rating': float_or_none(video.get('rating') or metas.get('rating')),
             'comment_count': int_or_none(video.get('commentsCount')),
         }
+
+
+class ADNSeasonIE(ADNIE):
```
Comment (pukkandan) [inline yt_dlp/extractor/adn.py:285]: Subclassing from another concrete extractor invites so many unexpected bugs. Create a `ADNBaseIE` with common methods from which both can be derived

File: README.md (hunk 5 above)
Comment (infanf) [reply README.md:1891]: Didn't know this was possible to be honest, then this is in deed not needed

**Comment by infanf** [comment]
German site is still in beta, but it works now (at least for some episodes). This means I was able to successfully run the tests locally.

File: yt_dlp/extractor/adn.py [hunk 7]
```diff
@@ -255,3 +290,42 @@ def _real_extract(self, url):
...
+                'order': 'asc',
+                'limit': '-1',
+            })
+        entries = []
+        for episode in episodes['videos']:
+            entries.append(self.url_result(
+                'https://animationdigitalnetwork.%s/video/%s/%s' % (lang, video_show_slug, episode['id']),
+                ie=ADNIE.ie_key(),
+                video_id=episode['id']
+            ))
```
Comment (seproDev) [inline yt_dlp/extractor/adn.py:330]:
```suggestion
        for episode_id in traverse_obj(episodes, ('videos', ..., 'id', {str})):
            entries.append(self.url_result(
                f'https://animationdigitalnetwork.{lang}/video/{video_show_slug}/{episode_id}',
                ADNIE, episode_id))
```
Add `from ..utils.traversal import traverse_obj` to the imports.

File: yt_dlp/extractor/adn.py [hunk 8]
```diff
@@ -255,3 +290,42 @@ def _real_extract(self, url):
...
+        return self.playlist_result(entries, show_id, show['title'])
```
Comment (seproDev) [inline yt_dlp/extractor/adn.py:331]: Title should be non fatal
```suggestion
        return self.playlist_result(entries, show_id, show.get('title'))
```

**Review by seproDev — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/adn.py (hunk 7 above)
Comment (infanf) [reply yt_dlp/extractor/adn.py]: didn't work with `{str}`

File: yt_dlp/extractor/adn.py [hunk 9]
```diff
@@ -255,3 +290,42 @@ def _real_extract(self, url):
...
+            video_base_url, video_show_slug,
+            'Downloading show JSON metadata',
+            headers=self._HEADERS)['show']
+        show_id = show['id']
+        lang = self._get_distribution_language(url)
+        episodes = self._download_json(
+            self._API_BASE_URL + 'video/show/%s' % show_id, video_show_slug,
+            'Downloading episode list', headers={
+                'X-Target-Distribution': lang,
+            }, query={
```
[the remark this answers is not in the document]
Comment (infanf) [reply yt_dlp/extractor/adn.py]: It works without it, but I guess it won't hurt to add the HEADERS. The website sends them with that request.

File: yt_dlp/extractor/adn.py (hunk 7 above)
Comment (seproDev) [reply yt_dlp/extractor/adn.py]: Oh, those get returned as integers. Use `{str_or_none}` from `..utils`.

Comment (infanf) [reply yt_dlp/extractor/adn.py]: Thanks, that works

File: yt_dlp/extractor/adn.py [hunk 10]
```diff
@@ -255,3 +288,38 @@ def _real_extract(self, url):
...
+        },
+        # 'skip': 'Only available in French end German speaking Europe',
+    }]
+
+    def _real_extract(self, url):
+        lang, video_show_slug = self._match_valid_url(url).group('lang', 'id')
+        show = self._download_json(
+            f'{self._API_BASE_URL}show/{video_show_slug}/', video_show_slug,
+            'Downloading show JSON metadata', headers=self._HEADERS)['show']
+        show_id = show['id']
```
Comment (seproDev) [inline yt_dlp/extractor/adn.py:310]:
```suggestion
        show_id = str(show['id'])
```

File: yt_dlp/extractor/adn.py [hunk 11]
```diff
@@ -255,3 +288,38 @@ def _real_extract(self, url):
...
         }
+
+
+class ADNSeasonIE(ADNBaseIE):
+    _VALID_URL = r'https?://(?:www\.)?(?:animation|anime)digitalnetwork\.(?P<lang>fr|de)/video/(?P<id>[^/]+)/?$'
+    _TESTS = [{
+        'url': 'https://animationdigitalnetwork.fr/video/tokyo-mew-mew-new',
+        'playlist_count': 12,
+        'info_dict': {
+            'id': 911,
```
Comment (seproDev) [inline yt_dlp/extractor/adn.py:299]:
```suggestion
            'id': '911',
```

File: yt_dlp/extractor/adn.py [hunk 12]
```diff
@@ -255,3 +288,38 @@ def _real_extract(self, url):
             'average_rating': float_or_none(video.get('rating') or metas.get('rating')),
             'comment_count': int_or_none(video.get('commentsCount')),
         }
+
+
+class ADNSeasonIE(ADNBaseIE):
+    _VALID_URL = r'https?://(?:www\.)?(?:animation|anime)digitalnetwork\.(?P<lang>fr|de)/video/(?P<id>[^/]+)/?$'
```
Comment (seproDev) [inline yt_dlp/extractor/adn.py:294]:
```suggestion
    _VALID_URL = r'https?://(?:www\.)?(?:animation|anime)digitalnetwork\.(?P<lang>fr|de)/video/(?P<id>[^/]+)/?(?:$|[#?])'
```

**Review by seproDev — APPROVED** [review]
For the future, please don't force push.

File: yt_dlp/extractor/adn.py [hunk 13]
```diff
@@ -19,15 +20,34 @@
...
+        'end': 3,
+    }
+    _LINE_ALIGN_MAP = {
+        'middle': 8,
+        'end': 4,
+    }
+
+
+class ADNIE(ADNBaseIE):
+    _VALID_URL = r'https?://(?:www\.)?(?:animation|anime)digitalnetwork\.(?P<lang>fr|de)/video/[^/]+/(?P<id>\d+)'
```
Comment (bashonly) [inline yt_dlp/extractor/adn.py:50]:
```suggestion
    _VALID_URL = r'https?://(?:www\.)?(?:animation|anime)digitalnetwork\.(?P<lang>fr|de)/video/[^/?#]+/(?P<id>\d+)'
```

File: yt_dlp/extractor/adn.py [hunk 14]
```diff
@@ -255,3 +288,38 @@ def _real_extract(self, url):
             'average_rating': float_or_none(video.get('rating') or metas.get('rating')),
             'comment_count': int_or_none(video.get('commentsCount')),
         }
+
+
+class ADNSeasonIE(ADNBaseIE):
+    _VALID_URL = r'https?://(?:www\.)?(?:animation|anime)digitalnetwork\.(?P<lang>fr|de)/video/(?P<id>[^/]+)/?(?:$|[#?])'
```
Comment (bashonly) [inline yt_dlp/extractor/adn.py:294]:
```suggestion
    _VALID_URL = r'https?://(?:www\.)?(?:animation|anime)digitalnetwork\.(?P<lang>fr|de)/video/(?P<id>[^/?#]+)/?(?:$|[#?])'
```

File: yt_dlp/extractor/adn.py [hunk 15]
```diff
@@ -255,3 +288,38 @@ def _real_extract(self, url):
...
+            }, query={
+                'order': 'asc',
+                'limit': '-1',
+            })
+        entries = []
+        for episode_id in traverse_obj(episodes, ('videos', ..., 'id', {str_or_none})):
+            entries.append(self.url_result(
+                f'https://animationdigitalnetwork.{lang}/video/{video_show_slug}/{episode_id}',
+                ADNIE, episode_id))
+        return self.playlist_result(entries, show_id, show.get('title'))
```
Comment (bashonly) [inline yt_dlp/extractor/adn.py:325]:
```suggestion
            })

        def entries():
            for episode_id in traverse_obj(episodes, ('videos', ..., 'id', {str_or_none})):
                yield self.url_result(
                    f'https://animationdigitalnetwork.{lang}/video/{video_show_slug}/{episode_id}',
                    ADNIE, episode_id)

        return self.playlist_result(entries(), show_id, show.get('title'))
```

File: yt_dlp/extractor/adn.py [hunk 16]
```diff
@@ -19,15 +20,34 @@
...
 
 
-class ADNIE(InfoExtractor):
+class ADNBaseIE(InfoExtractor):
     IE_DESC = 'Animation Digital Network'
-    _VALID_URL = r'https?://(?:www\.)?(?:animation|anime)digitalnetwork\.fr/video/[^/]+/(?P<id>\d+)'
+    _NETRC_MACHINE = 'animationdigitalnetwork'
+    _BASE = 'animationdigitalnetwork.fr'
+    _API_BASE_URL = 'https://gw.api.' + _BASE + '/'
+    _PLAYER_BASE_URL = _API_BASE_URL + 'player/'
```
Comment (bashonly) [inline yt_dlp/extractor/adn.py:35]:
```suggestion
    _API_BASE_URL = f'https://gw.api.{_BASE}/'
    _PLAYER_BASE_URL = f'{_API_BASE_URL}player/'
```

File: yt_dlp/extractor/adn.py [hunk 17]
```diff
@@ -19,15 +20,34 @@
     long_to_bytes,
     pkcs1pad,
     strip_or_none,
+    str_or_none,
     try_get,
     unified_strdate,
     urlencode_postdata,
 )
```
Comment (bashonly) [inline yt_dlp/extractor/adn.py:29]:
```suggestion
)
from ..utils.traversal import traverse_obj

```

File: yt_dlp/extractor/adn.py [hunk 18]
```diff
@@ -8,6 +8,7 @@
 from ..aes import aes_cbc_decrypt_bytes, unpad_pkcs7
 from ..compat import compat_b64decode
 from ..networking.exceptions import HTTPError
+from ..utils.traversal import traverse_obj
```
Comment (bashonly) [inline yt_dlp/extractor/adn.py:11]:
```suggestion
```

**Review by bashonly — APPROVED** [review]
